File: doubly_linked_list.py
class Node(object):
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class Deque(object):
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None

    def push_right(self, val):
        new_node = Node(val)
        self.size += 1
        if not self.tail:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

    def pop_right(self):
        if not self.tail:
            raise IndexError("Cannot pop from empty deque")
        else:
            val = self.tail.val
            self.size -= 1

            if self.head == self.tail:
                self.head = None
                self.tail = None

            else:
                self.tail = self.tail.prev
                self.tail.next = None

            return val

    def push_left(self, val):
        new_node = Node(val)
        self.size += 1

        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

    def pop_left(self):
        if not self.head:
            raise IndexError("Cannot pop from empty deque")
        else:
            val = self.head.val
            self.size -= 1

            if self.head == self.tail:
                self.head = None
                self.tail = None
            else:
                self.head = self.head.next
                self.head.prev = None

            return val

File: dsa/doubly_linked_list.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class Deque {
  constructor() {
    this.size = 0
    this.head = null
    this.tail = null
  } 

  pushRight(val) {
    const newNode = new Node(val)
    this.size++ 

    if (!this.head) {
      this.head = newNode
      this.tail = newNode
    } else {
      newNode.prev = this.tail
      this.tail.next = newNode
      this.tail = newNode
    }
  }

  pushLeft(val) {
    const newNode = new Node(val)
    this.size++

    if (!this.head) {
      this.head = newNode
      this.tail = newNode
    } else {
      newNode.next = this.head
      this.head.prev = newNode
      this.head = newNode
    }
  }

  popRight() {
    if (!this.head) {
      throw Error("Should not be empty")
    }
    if (this.head === this.tail) {
      this.head = null
      this.tail = null
    } else {
      const val = this.tail.val
      this.size--

      this.tail = this.tail.prev
      this.tail.next = null

      return val
    }
  }

  popLeft() {
    if (!this.head) {
      throw Error("Should not be empty")
    }
    if (this.head === this.tail) {
      this.head = null
      this.tial = null
    } else {
      const val = this.head.val
      this.size--

      this.head = this.head.next
      this.head.prev = null

      return val
    }
  }
}

const deque = new Deque()

// Test pushRight
deque.pushRight(1)
deque.pushRight(2)
deque.pushRight(3)
console.assert(deque.size === 3, "pushRight: size should be 3")
console.assert(deque.tail.val === 3, "pushRight: tail value should be 3")
console.assert(deque.head.val === 1, "pushRight: head value should be 1")

// Test popRight 
const rightVal = deque.popRight()
console.assert(rightVal === 3, "popRight: should return 3")
console.assert(deque.size === 2, "popRight: size should be 2")
console.assert(deque.tail.val === 2, "popRight: new tail should be 2")

// Test pushLeft
deque.pushLeft(0)
console.assert(deque.size === 3, "pushLeft: size should be 3") 
console.assert(deque.head.val === 0, "pushLeft: head value should be 0")

// Test popLeft
const leftVal = deque.popLeft()
console.assert(leftVal === 0, "popLeft: should return 0")
console.assert(deque.size === 2, "popLeft: size should be 2")
console.assert(deque.head.val === 1, "popLeft: new head should be 1")

console.log("\x1b[32m%s\x1b[0m", "All tests passed!")
